### ISSUE TITLE
Integrate RL training pipeline with CausalMoMa for whole-body mobile manipulation

### ISSUE THREAD

**Issue by H-tr** [Robot-TLab/TyGrit#9]
## Summary

Add an RL training pipeline for learning whole-body mobile manipulation ("grasp from anywhere") using the [CausalMoMa](https://arxiv.org/abs/2305.04866) algorithm (Causal Policy Gradient, RSS 2023). The RL pipeline reuses the existing `ManiSkillFetchRobot` interface and leverages ManiSkill3's GPU-parallel simulation for training throughput.

## Motivation

The current system is fully model-based: SubGoalGenerator → MotionPlanner (VAMP) → MPC Controller. CausalMoMa offers a learned alternative that:

- Automatically discovers causal dependencies between action subsets (base, arm, head, gripper) and reward terms (reach, gaze, grasp)
- Reduces policy gradient variance by applying per-term advantages only to causally-related action subsets
- Enables whole-body coordination without manual task decomposition
- Demonstrated sim-to-real transfer on Fetch and HSR robots

### Two-layer architecture

```
                    ┌────────────────────────────────┐
                    │   FetchGraspEnv (gymnasium.Env) │  ← task layer
                    │   reward + termination + obs    │
                    └──────────────┬─────────────────┘
                                   │ wraps
                        ┌──────────┼──────────┐
                        │          │          │
               ManiSkillFetch  ManiSkillFetch  ROSFetch      ← robot layer
               Robot (planning) RobotVec (RL)  Robot (real)
                        │          │          │
                        └──── FetchRobot ─────┘
                                   │
                              RobotBase (protocol)
```

**Layer 1 — Robot** (`RobotBase` / `FetchRobot`): shared interface for all backends. The RL env reuses the same abstract robot class so trained policies are testable on any backend (single-env ManiSkill, ROS, etc.) without modification.

**Layer 2 — Task** (`FetchGraspEnv`): a `gymnasium.Env` that wraps any `FetchRobot`, adding reward computation, episode management, and observation encoding.

### Key insight: `ManiSkillFetchRobot` is NOT planning-coupled

The existing env is general-purpose robot control. The action format `[v, w, torso_vel, 7 arm_vels]` works for any velocity controller — MPC or learned policy. Head control via `look_at()` and gripper via `control_gripper()` are standard `RobotBase` methods. Trajectory execution methods (`execute_trajectory`) are simply not called by the RL pipeline.

The **only issue** with `num_envs > 1` is single-env assumptions in observation parsing (hardcoded `[0]` batch indexing, `float()` calls, returning single `SensorSnapshot`). Solution: subclass that overrides only the batch-dim handling methods.

### Reusable existing code

| Module | Used by RL for |
|--------|----------------|
| `kinematics/fetch/fk_torch.py` | `batch_forward_kinematics()` — GPU-batched EE position for reward |
| `kinematics/fetch/constants.py` | Joint names/limits, camera offsets (obs normalization, gaze reward) |
| `envs/fetch/maniskill.py` | Base class — action slices, joint maps, base offset calibration, head PD, obs parsing logic |
| `utils/tensor.py` | `to_numpy()` conversion |

### Action space (CausalMoMa factoring)

The policy outputs the same 10-dim action that `ManiSkillFetchRobot.step()` accepts: `[v, w, torso_vel, 7 arm_vels]`. Head is driven through `look_at()` → PD controller, gripper through `control_gripper()`.

For CausalMoMa's causal matrix, the action subsets are:
- `a_base = [v, w]` (indices 0:2)
- `a_arm = [torso_vel, 7 arm_vels]` (indices 2:10)

Head and gripper are controlled via `RobotBase` methods, not the action vector — they participate in the causal structure through the task env's `step()` which calls `look_at()` and `control_gripper()` before `robot.step()`.

### Factored reward (for CausalMoMa per-term advantages)

```
R = w_reach * r_reach       # -||ee_pos - target_pos||
  + w_gaze  * r_gaze        # target in camera frustum
  + w_grasp * r_grasp       # object grasped + lifted
  + w_smooth * r_smooth     # action penalty
```

Individual `r_j` terms are returned in `info["reward_terms"]` for CausalMoMa's causal policy gradient.

### Phase 1: Vectorized robot backend

- [ ] Add `num_envs` field to `FetchEnvConfig` (or new `FetchRLEnvConfig`)
- [ ] Create `ManiSkillFetchRobotVec(ManiSkillFetchRobot)` in `envs/fetch/maniskill_vec.py`
  - Override `__init__` to pass `num_envs` to `gym.make()`
  - Override `_parse_observation()` for batched `(N, ...)` tensors
  - Override `step()` for batched actions `(N, 10)` and batched returns
  - Override `_init_qpos_world_offset()` for batched base calibration
  - Inherit: action slices, joint name map, `look_at()`, head PD, `control_gripper()`, `_assemble_action()`
- [ ] Verify `num_envs=1` mode matches existing `ManiSkillFetchRobot` behavior exactly

### Phase 2: Task environment

- [ ] Create `FetchGraspEnv(gymnasium.Env)` in `envs/fetch/task.py`
  - Wraps any `FetchRobot` (single or vectorized)
  - `reset()`: randomize robot config + object placement, return encoded obs
  - `step()`: decode policy action → `look_at()` + `control_gripper()` + `robot.step()`, compute reward, check termination
  - Observation: `[base(3), torso(1), arm(7), head(2), gripper(1), ee_pos(3), target_pos(3)]` — all derivable from `RobotState` + FK
  - Episode termination: object grasped + lifted, or timeout
- [ ] Create `FetchGraspConfig` in `envs/fetch/rl_config.py` (reward weights, max_episode_steps, randomization ranges)
- [ ] Verify policy portability: train on `ManiSkillFetchRobotVec`, evaluate on `ManiSkillFetchRobot`

### Phase 3: Reward and observation functions

- [ ] Create `rl/rewards.py` — pure functions operating on batched tensors
  - `reach_reward(ee_pos, target_pos) → (B,)` — uses `batch_forward_kinematics()`
  - `gaze_reward(target_pos, camera_pose, intrinsics) → (B,)` — target in frustum
  - `grasp_reward(gripper_state, object_lifted) → (B,)` — sparse success
  - `action_penalty(action) → (B,)` — smoothness
- [ ] Create `rl/obs.py` — `encode_obs(RobotState, target_pos) → ndarray` (pure function, shared between training and eval)
- [ ] Unit tests for reward functions

### Phase 4: CausalMoMa training

- [ ] Create `rl/causal.py` — causal discovery
  - Collect random rollouts
  - Estimate conditional mutual information (CMI) between action subsets and reward terms
  - Threshold → binary causal matrix `B` of shape `(K, J)`
- [ ] Create `rl/policy.py` — factored policy network
  - Shared encoder → separate action heads per subset (`π_base`, `π_arm`)
  - Compatible with per-subset advantage signals
- [ ] Create `rl/config.py` — training hyperparams (algorithm, learning rate, num_envs, epochs, etc.)
- [ ] Create `rl/train.py` — modified PPO with causal policy gradient
  - Per-reward-term advantage computation
  - Aggregate advantages via causal matrix B for each action subset
  - Standard PPO loss but with factored advantages

### Phase 5: Evaluation and baselines

- [ ] Evaluation script: load trained policy, run on single-env `ManiSkillFetchRobot`
- [ ] Compare against classical pipeline (existing grasp benchmark)
- [ ] Success rate, episode length, reward curves

## File Layout

```
TyGrit/
├── envs/fetch/
│   ├── maniskill.py          # existing (unchanged)
│   ├── maniskill_vec.py      # NEW: ManiSkillFetchRobotVec — batched for training
│   ├── task.py               # NEW: FetchGraspEnv(gymnasium.Env) — wraps any FetchRobot
│   ├── rl_config.py          # NEW: FetchGraspConfig (reward weights, episode params)
│   ├── config.py             # existing (may add num_envs field)
│   ├── fetch.py              # existing (update factory for new backend)
│   └── ros.py                # existing (unchanged)
├── rl/                       # NEW
│   ├── rewards.py            # pure reward functions (batched tensors)
│   ├── obs.py                # observation encoding (pure function)
│   ├── causal.py             # CMI → causal matrix B
│   ├── policy.py             # factored policy network
│   ├── config.py             # training hyperparams
│   └── train.py              # entry point
└── (everything else unchanged)
```

## Design Constraints

- Shared `RobotBase` / `FetchRobot` interface — trained policy must be testable on any backend
- Pure functions for rewards and obs encoding (matches project convention)
- Frozen dataclass configs
- No changes to existing modules — RL pipeline is purely additive
- GPU-parallel training via ManiSkill3 `num_envs`, single-env eval via standard `RobotBase`

## References

- [CausalMoMa paper (RSS 2023)](https://arxiv.org/abs/2305.04866)
- [CausalMoMa code](https://github.com/JiahengHu/CausalMoMa)
- [ManiSkill3 GPU simulation](https://maniskill.readthedocs.io/en/latest/user_guide/concepts/gpu_simulation.html)
